File: hexameter-core/src/test/java/org/codetome/hexameter/core/api/defaults/DefaultHexagonDataStorageTest.java
package org.codetome.hexameter.core.api.defaults;

import org.codetome.hexameter.core.api.CubeCoordinate;
import org.codetome.hexameter.core.backport.Optional;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.MockitoAnnotations.initMocks;

public class DefaultHexagonDataStorageTest {

    private static final CubeCoordinate TEST_CUBE_COORDINATE = CubeCoordinate.fromCoordinates(1, 1);
    private static final CubeCoordinate ANOTHER_TEST_CUBE_COORDINATE = CubeCoordinate.fromCoordinates(2, 1);

    private DefaultHexagonDataStorage target;

    @Mock
    private DefaultSatelliteData defaultSatelliteDataMock;

    @Before
    public void setUp() {
        initMocks(this);
        target = new DefaultHexagonDataStorage();
    }

    @Test
    public void shouldAddCoordinateWhenAddCoordinateIsCalledWithCubeCoordinate() {
        target.addCoordinate(TEST_CUBE_COORDINATE);
        assertThat(target.containsCoordinate(TEST_CUBE_COORDINATE)).isTrue();
    }

    @Test
    public void shouldAddCoordinateAndDataWhenCalledWithBoth() {
        target.addCoordinate(TEST_CUBE_COORDINATE, defaultSatelliteDataMock);
        assertThat(target.containsCoordinate(TEST_CUBE_COORDINATE)).isTrue();
        assertThat(target.hasDataFor(TEST_CUBE_COORDINATE));
        assertThat(target.getSatelliteDataBy(TEST_CUBE_COORDINATE).get()).isEqualTo(defaultSatelliteDataMock);
    }

    @Test
    public void shouldProperlySignalModificationWhenReplacesPreviousValue() {
        target.addCoordinate(TEST_CUBE_COORDINATE);
        final boolean result = target.addCoordinate(TEST_CUBE_COORDINATE, defaultSatelliteDataMock);
        assertThat(result).isTrue();
    }

    @Test
    public void shouldProperlyReturnEmptyOptionalWhenTryingToGetNonPresentData() {
        target.addCoordinate(TEST_CUBE_COORDINATE);
        final Optional<DefaultSatelliteData> result = target.getSatelliteDataBy(TEST_CUBE_COORDINATE);
        assertThat(result.isPresent()).isFalse();
    }

    @Test
    public void shouldProperlyReturnEmptyOptionalWhenTryingToGetDataWithNonPresentKey() {
        final Optional<DefaultSatelliteData> result = target.getSatelliteDataBy(TEST_CUBE_COORDINATE);
        assertThat(result.isPresent()).isFalse();
    }

    @Test
    public void shouldProperlyReturnFalseWhenContainsCoordinateIsCalledWithNonPresentCoordinate() {
        assertThat(target.containsCoordinate(TEST_CUBE_COORDINATE)).isFalse();
    }

    @Test
    public void shouldProperlyReturnTrueWhenContainsCoordinateIsCalledWithPresentCoordinate() {
        target.addCoordinate(TEST_CUBE_COORDINATE, defaultSatelliteDataMock);
        assertThat(target.containsCoordinate(TEST_CUBE_COORDINATE)).isTrue();
    }

    @Test
    public void shouldProperlyReturnFalseWhenHasDataForIsCalledAndThereIsNoData() {
        assertThat(target.hasDataFor(TEST_CUBE_COORDINATE)).isFalse();
    }

    @Test
    public void shouldProperlyReturnTrueWhenHasDataForIsCalledAndThereIsData() {
        target.addCoordinate(TEST_CUBE_COORDINATE, defaultSatelliteDataMock);
        assertThat(target.hasDataFor(TEST_CUBE_COORDINATE)).isTrue();
    }

    @Test
    public void shouldProperlyReturnCoordinatesWhenThereAreNonePresent() {
        assertThat(target.getCoordinates().count().toBlocking().single()).isEqualTo(0);
    }

    @Test
    public void shouldProperlyReturnCoordinatesWhenThereAreSeveralPresent() {
        target.addCoordinate(TEST_CUBE_COORDINATE, defaultSatelliteDataMock);
        target.addCoordinate(TEST_CUBE_COORDINATE, defaultSatelliteDataMock);
        target.addCoordinate(ANOTHER_TEST_CUBE_COORDINATE, defaultSatelliteDataMock);
        assertThat(target.getCoordinates().count().toBlocking().single()).isEqualTo(2);
    }

    @Test
    public void shouldProperlyClearDataForWhenCalledAndDataIsPresent() {
        target.addCoordinate(TEST_CUBE_COORDINATE, defaultSatelliteDataMock);
        assertThat(target.hasDataFor(TEST_CUBE_COORDINATE)).isTrue();
        target.clearDataFor(TEST_CUBE_COORDINATE);
        assertThat(target.hasDataFor(TEST_CUBE_COORDINATE)).isFalse();
    }

}

File: mixite2.core/src/commonTest/kotlin/org/hexworks/mixite2/core/api/defaults/DefaultHexagonDataStorageTest.kt
package org.hexworks.mixite2.core.api.defaults

import org.hexworks.mixite2.core.api.CubeCoordinate
import kotlin.test.*

class DefaultHexagonDataStorageTest {

    lateinit var target: DefaultHexagonDataStorage<DefaultSatelliteData>
    lateinit var defaultSatelliteDataStub: DefaultSatelliteData

    @BeforeTest
    fun setUp() {
        target = DefaultHexagonDataStorage()
        defaultSatelliteDataStub = DefaultSatelliteData()
    }

    @Test
    fun shouldAddCoordinateWhenAddCoordinateIsCalledWithCubeCoordinate() {
        target.addCoordinate(TEST_CUBE_COORDINATE)
        assertTrue {
            target.containsCoordinate(TEST_CUBE_COORDINATE)
        }
    }

    @Test
    fun shouldAddCoordinateAndDataWhenCalledWithBoth() {
        target.addCoordinate(TEST_CUBE_COORDINATE, defaultSatelliteDataStub)
        assertTrue {
            target.containsCoordinate(TEST_CUBE_COORDINATE)
        }
        assertTrue {
            target.hasDataFor(TEST_CUBE_COORDINATE)
        }
        assertEquals(expected = defaultSatelliteDataStub, actual = target.getSatelliteDataBy(TEST_CUBE_COORDINATE).get())
    }

    @Test
    fun shouldProperlySignalModificationWhenReplacesPreviousValue() {
        target.addCoordinate(TEST_CUBE_COORDINATE)

        assertTrue {
            target.addCoordinate(TEST_CUBE_COORDINATE, defaultSatelliteDataStub)
        }
    }

    @Test
    fun shouldProperlyReturnEmptyOptionalWhenTryingToGetNonPresentData() {
        target.addCoordinate(TEST_CUBE_COORDINATE)
        val result = target.getSatelliteDataBy(TEST_CUBE_COORDINATE)

        assertFalse {
            result.isPresent
        }
    }

    @Test
    fun shouldProperlyReturnEmptyOptionalWhenTryingToGetDataWithNonPresentKey() {
        val result = target.getSatelliteDataBy(TEST_CUBE_COORDINATE)

        assertFalse {
            result.isPresent
        }
    }

    @Test
    fun shouldProperlyReturnFalseWhenContainsCoordinateIsCalledWithNonPresentCoordinate() {
        assertFalse {
            target.containsCoordinate(TEST_CUBE_COORDINATE)
        }
    }

    @Test
    fun shouldProperlyReturnTrueWhenContainsCoordinateIsCalledWithPresentCoordinate() {
        target.addCoordinate(TEST_CUBE_COORDINATE, defaultSatelliteDataStub)

        assertTrue {
            target.containsCoordinate(TEST_CUBE_COORDINATE)
        }
    }

    @Test
    fun shouldProperlyReturnFalseWhenHasDataForIsCalledAndThereIsNoData() {
        assertFalse {
            target.hasDataFor(TEST_CUBE_COORDINATE)
        }
    }

    @Test
    fun shouldProperlyReturnTrueWhenHasDataForIsCalledAndThereIsData() {
        target.addCoordinate(TEST_CUBE_COORDINATE, defaultSatelliteDataStub)

        assertTrue {
            target.hasDataFor(TEST_CUBE_COORDINATE)
        }
    }

    @Test
    fun shouldProperlyReturnCoordinatesWhenThereAreNonePresent() {
        assertFalse {
            target.coordinates.iterator().hasNext()
        }
    }

    @Test
    fun shouldProperlyReturnCoordinatesWhenThereAreSeveralPresent() {
        target.addCoordinate(TEST_CUBE_COORDINATE, defaultSatelliteDataStub)
        target.addCoordinate(TEST_CUBE_COORDINATE, defaultSatelliteDataStub)
        target.addCoordinate(ANOTHER_TEST_CUBE_COORDINATE, defaultSatelliteDataStub)

        val iter = target.coordinates.iterator()
        var count = 0
        while (iter.hasNext()) {
            iter.next()
            count++
        }
        assertEquals(expected = 2, actual = count)
    }

    @Test
    fun shouldProperlyClearDataForWhenCalledAndDataIsPresent() {
        target.addCoordinate(TEST_CUBE_COORDINATE, defaultSatelliteDataStub)

        assertTrue {
            target.hasDataFor(TEST_CUBE_COORDINATE)
        }

        target.clearDataFor(TEST_CUBE_COORDINATE)

        assertFalse {
            target.hasDataFor(TEST_CUBE_COORDINATE)
        }
    }

    companion object {

        private val TEST_CUBE_COORDINATE = org.hexworks.mixite2.core.api.CubeCoordinate.fromCoordinates(1, 1)
        private val ANOTHER_TEST_CUBE_COORDINATE = org.hexworks.mixite2.core.api.CubeCoordinate.fromCoordinates(2, 1)
    }

}
